akka-stream/src/main/scala/akka/stream/impl/fusing/Interpreter.scala

```diff
@@ -238,22 +245,64 @@ private[akka] class OneBoundedInterpreter(ops: Seq[Stage[_, _]], val forkLimit:
      */
     def run(): Unit
 
+    /**
+     * This method shall return true iff this is a push-style state.
+     */
+    def comingFromUpstream: Boolean
+
+    /**
+     * This method shall return true iff this is a pull-style state.
+     */
+    def comingFromDownstream: Boolean
+
+    /**
+     * This method shall return true iff this is an async entry (“from the side”).
+     */
+    def comingFromSide: Boolean
+
     override def push(elem: Any): DownstreamDirective = {
       ReactiveStreamsCompliance.requireNonNullElement(elem)
-      if (currentOp.holding) throw new IllegalStateException("Cannot push while holding, only pushAndPull")
-      currentOp.allowedToPush = false
+      if (currentOp.isDetached) {
+        if (comingFromUpstream)
+          throw new IllegalStateException("Cannot push during onPush, only pull or pushAndPull")
+        if (comingFromSide && !currentOp.holdingDownstream)
+          throw new IllegalStateException("Cannot push during onAsyncInput unless isHoldingDownstream")
+      }
+      currentOp.precedingWasPull = false
+      currentOp.holdingDownstream = false
       elementInFlight = elem
       state = Pushing
       null
     }
 
     override def pull(): UpstreamDirective = {
-      if (currentOp.holding) throw new IllegalStateException("Cannot pull while holding, only pushAndPull")
-      currentOp.allowedToPush = !currentOp.isInstanceOf[DetachedStage[_, _]]
+      if (currentOp.isDetached) {
+        if (comingFromDownstream)
+          throw new IllegalStateException("Cannot pull during onPull, only push or pushAndPull")
+        if (comingFromSide && !currentOp.holdingUpstream)
+          throw new IllegalStateException("Cannot push during onAsyncInput unless isHoldingUpstream")
+      }
+      currentOp.precedingWasPull = true
+      currentOp.holdingUpstream = false
       state = Pulling
       null
     }
 
+    override def getAsyncCallback(): AsyncCallback[Any] = {
+      val current = currentOp.asInstanceOf[AsyncStage[Any, Any, Any]]
+      val context = current.context // avoid concurrent access (to avoid @volatile)
+      new AsyncCallback[Any] {
```

Comment:
@rkuhn yes ,for the ```if (elemInFlight != null)  ```,could not click the line on phone.I know the result is identical,but you are compare the reference,maybe I have been followed the book too much:P.